File: sdks/eatf-verifier-ts/src/tsa.ts
/**
 * Phase 1 step 1.20 v0.2 — RFC 3161 timestamp parsing + verification.
 *
 * <p>This module parses a base64-encoded TimeStampToken via {@code pkijs}
 * and performs four independent checks:
 *
 * <ol>
 *   <li><b>tsaPresent</b> — the token is non-empty and parses as
 *       {@code ContentInfo} with the CMS SignedData OID
 *       ({@code 1.2.840.113549.1.7.2}).</li>
 *   <li><b>messageImprintMatches</b> — the {@code TSTInfo.messageImprint
 *       .hashedMessage} octet string equals {@code SHA-256(expectedHashHex
 *       as ASCII)}. The EATF AEP profile (see {@code docs/specs/aep
 *       -profile-v1.md} §7) hashes the lower-case hex string of the
 *       canonical SHA-256, not the canonical bytes themselves; this is
 *       the deliberate compatibility point with Java's
 *       {@code RealTsaServiceImpl}.</li>
 *   <li><b>signatureVerified</b> — the embedded
 *       {@code SignerInfo.signature} validates against the embedded
 *       signing certificate's public key over the
 *       {@code signedAttributes} (CMS RFC 5652 §5.4). This is the
 *       "internally consistent" check: the entity holding the
 *       certificate did sign this messageImprint at this genTime.</li>
 *   <li><b>genTime</b> — the {@code TSTInfo.genTime} timestamp, returned
 *       as a JavaScript {@code Date} for downstream policy. Not a check
 *       in itself; the caller decides whether to enforce a window.</li>
 * </ol>
 *
 * <p>Trust-anchor chain validation is a <em>separate</em> step exposed
 * by {@link verifyTsaTrust} so an operator can opt in / out
 * independently of the internal-consistency checks.
 *
 * <p>v0.1-alpha shipped a regex-over-hex heuristic for
 * messageImprintMatches and left signatureVerified always {@code null}.
 * This module replaces both with deterministic ASN.1 parsing and a
 * real Web Crypto signature verify.
 */

import * as asn1js from "asn1js";
import {
  Certificate,
  ContentInfo,
  SignedData,
  setEngine,
  CryptoEngine,
} from "pkijs";

import { sha256 } from "./hash.js";
import { decodeBase64 } from "./rsa.js";
import type { TsaTrustResult } from "./tsa-trust-list.js";

// PKIjs requires an explicit crypto engine. In Node 20+ and modern
// browsers `globalThis.crypto.subtle` exists; we register it once
// per module load and keep the registration idempotent across hot
// reload + test runners.
let engineConfigured = false;
function configureEngine(): void {
  if (engineConfigured) return;
  const subtle = globalThis.crypto?.subtle;
  if (!subtle) {
    throw new Error(
      "@eatf/verifier requires globalThis.crypto.subtle (Node 20+ or modern browser).",
    );
  }
  setEngine(
    "@eatf/verifier-engine",
    new CryptoEngine({ name: "@eatf/verifier-engine", crypto: globalThis.crypto, subtle }),
  );
  engineConfigured = true;
}

export type TsaCheck = {
  /** Token is non-empty and parses as RFC 3161 ContentInfo + SignedData. */
  tsaPresent: boolean;
  /** Imprint inside the token equals SHA-256 of the expected hash hex. */
  messageImprintMatches: boolean | null;
  /** SignerInfo signature verifies against the embedded cert's public key. */
  signatureVerified: boolean | null;
  /** Hash algorithm OID inside the imprint (typically 2.16.840.1.101.3.4.2.1 for SHA-256). */
  imprintAlgorithmOid: string | null;
  /** TSTInfo.genTime, as a JavaScript Date. */
  genTime: Date | null;
  /** Number of certificates embedded in the SignedData.certificates field. */
  embeddedCertCount: number;
  /** Subject DN of the embedded signing cert (informational). */
  signerSubject: string | null;
  /** Issuer DN of the embedded signing cert (informational). */
  signerIssuer: string | null;
  /** Total bytes of the DER-decoded token. */
  rawSizeBytes: number;
  /** Human-readable note when a step short-circuits. */
  note?: string;
};

const SHA256_OID = "2.16.840.1.101.3.4.2.1";
const CMS_SIGNED_DATA_OID = "1.2.840.113549.1.7.2";

function emptyCheck(note: string, rawSizeBytes = 0): TsaCheck {
  return {
    tsaPresent: false,
    messageImprintMatches: null,
    signatureVerified: null,
    imprintAlgorithmOid: null,
    genTime: null,
    embeddedCertCount: 0,
    signerSubject: null,
    signerIssuer: null,
    rawSizeBytes,
    note,
  };
}

export async function inspectTsa(
  timestampBase64: string,
  expectedHashHex: string,
): Promise<TsaCheck> {
  if (!timestampBase64 || !timestampBase64.trim()) {
    return emptyCheck("No timestamp.tsr in package.");
  }

  let der: Uint8Array;
  try {
    der = decodeBase64(timestampBase64.trim());
  } catch (e) {
    return emptyCheck(`timestamp.tsr is not valid base64: ${describeError(e)}`);
  }

  if (der.length < 32) {
    return emptyCheck("TSR too short to be a TimeStampToken.", der.length);
  }

  configureEngine();

  // ----- Parse outer ContentInfo -----
  let contentInfo: ContentInfo;
  try {
    const fresh = der.slice();
    const asn = asn1js.fromBER(fresh.buffer);
    if (asn.offset === -1) {
      return emptyCheck("Outer DER did not parse as ASN.1.", der.length);
    }
    contentInfo = new ContentInfo({ schema: asn.result });
  } catch (e) {
    return emptyCheck(`ContentInfo parse failed: ${describeError(e)}`, der.length);
  }

  if (contentInfo.contentType !== CMS_SIGNED_DATA_OID) {
    return emptyCheck(
      `Unexpected ContentInfo.contentType ${contentInfo.contentType} (want CMS SignedData ${CMS_SIGNED_DATA_OID}).`,
      der.length,
    );
  }

  // ----- Parse SignedData + extract TSTInfo eContent bytes -----
  let signedData: SignedData;
  let tstInfoBytes: Uint8Array;
  try {
    signedData = new SignedData({ schema: contentInfo.content });
    const eContent = signedData.encapContentInfo.eContent;
    if (!eContent) {
      return emptyCheck("SignedData.encapContentInfo has no eContent (detached?).", der.length);
    }
    const vb = eContent.valueBlock as unknown as {
      valueHexView?: Uint8Array;
      valueHex?: ArrayBuffer;
    };
    if (vb.valueHexView) {
      tstInfoBytes = vb.valueHexView;
    } else if (vb.valueHex) {
      tstInfoBytes = new Uint8Array(vb.valueHex);
    } else {
      return emptyCheck("Cannot extract eContent bytes from SignedData.", der.length);
    }
  } catch (e) {
    return emptyCheck(`SignedData parse failed: ${describeError(e)}`, der.length);
  }

  // ----- Extract TSTInfo fields -----
  let imprintHashBytes: Uint8Array | null = null;
  let imprintAlgorithmOid: string | null = null;
  let genTime: Date | null = null;
  try {
    const tstAsn = asn1js.fromBER(tstInfoBytes.slice().buffer);
    if (tstAsn.offset === -1) {
      return emptyCheck("TSTInfo did not parse as ASN.1.", der.length);
    }
    // TSTInfo ::= SEQUENCE {
    //   version            INTEGER,
    //   policy             OBJECT IDENTIFIER,
    //   messageImprint     MessageImprint,
    //   serialNumber       INTEGER,
    //   genTime            GeneralizedTime,
    //   ...optionals
    // }
    const tstSeq = tstAsn.result as asn1js.Sequence;
    const items = tstSeq.valueBlock.value;
    // messageImprint is the 3rd element (index 2): SEQUENCE { algId, octetString }
    const mi = items[2] as asn1js.Sequence;
    const algId = mi.valueBlock.value[0] as asn1js.Sequence;
    const oidObj = algId.valueBlock.value[0] as asn1js.ObjectIdentifier;
    imprintAlgorithmOid = oidObj.valueBlock.toString();
    const hashOctet = mi.valueBlock.value[1] as asn1js.OctetString;
    const hashView = (hashOctet.valueBlock as unknown as {
      valueHexView?: Uint8Array;
      valueHex?: ArrayBuffer;
    });
    if (hashView.valueHexView) {
      imprintHashBytes = new Uint8Array(hashView.valueHexView);
    } else if (hashView.valueHex) {
      imprintHashBytes = new Uint8Array(hashView.valueHex);
    }
    // genTime is index 4 (GeneralizedTime).
    const gtNode = items[4] as asn1js.GeneralizedTime;
    if (gtNode && typeof gtNode.toDate === "function") {
      genTime = gtNode.toDate();
    }
  } catch (e) {
    return emptyCheck(
      `TSTInfo field extraction failed: ${describeError(e)}`,
      der.length,
    );
  }

  // ----- Compare imprint to expected -----
  let messageImprintMatches: boolean | null = null;
  if (imprintHashBytes && imprintAlgorithmOid === SHA256_OID) {
    const tsaInput = new TextEncoder().encode(expectedHashHex);
    const expectedDigest = new Uint8Array(await sha256(tsaInput));
    messageImprintMatches = constantTimeEqual(imprintHashBytes, expectedDigest);
  }

  // ----- Extract signing cert + verify SignerInfo signature -----
  const certs = (signedData.certificates ?? []).filter(
    (c): c is Certificate => c instanceof Certificate,
  );
  let signerSubject: string | null = null;
  let signerIssuer: string | null = null;
  if (certs.length > 0) {
    signerSubject = describeDn(certs[0]!.subject.typesAndValues as DnTypeAndValue[]);
    signerIssuer = describeDn(certs[0]!.issuer.typesAndValues as DnTypeAndValue[]);
  }

  let signatureVerified: boolean | null = null;
  try {
    if (certs.length === 0) {
      signatureVerified = null;
    } else {
      // pkijs.SignedData.verify resolves to a boolean (true = OK,
      // false = signature mismatch). It throws on structural problems
      // (missing signed attributes, OID mismatches, etc.) which we
      // surface separately below.
      const ok = await signedData.verify({
        signer: 0,
        data: tstInfoBytes.slice().buffer,
        trustedCerts: certs,
      });
      signatureVerified = ok === true;
    }
  } catch (e) {
    signatureVerified = false;
    return {
      tsaPresent: true,
      messageImprintMatches,
      signatureVerified,
      imprintAlgorithmOid,
      genTime,
      embeddedCertCount: certs.length,
      signerSubject,
      signerIssuer,
      rawSizeBytes: der.length,
      note: `SignerInfo verify threw: ${describeError(e)}`,
    };
  }

  return {
    tsaPresent: true,
    messageImprintMatches,
    signatureVerified,
    imprintAlgorithmOid,
    genTime,
    embeddedCertCount: certs.length,
    signerSubject,
    signerIssuer,
    rawSizeBytes: der.length,
    note:
      signatureVerified === true
        ? "TSA signature verified against embedded cert. Trust-anchor chain validation is a separate step (verifyTsaTrust)."
        : signatureVerified === false
        ? "TSA signature did not verify against embedded cert. Treat as tampered."
        : "TSA signature not checked (no embedded cert).",
  };
}

/**
 * Cross-check the embedded signing cert chain against a pinned trust
 * list. This is independent of {@link inspectTsa}; an operator may
 * choose to require it (production) or skip it (dev / no trust anchors
 * pinned yet).
 *
 * <p>v0.2 does a minimal validation: the embedded signer's issuer DN
 * must match the subject DN of at least one provided trust-list root.
 * Full RFC 5280 path validation (NotBefore / NotAfter / KeyUsage /
 * EKU / Basic Constraints / CRL / OCSP) is v0.3 territory.
 *
 * @param check the result of {@link inspectTsa}
 * @param trustList PEM-encoded root certificates. Empty → skipped.
 */
export async function verifyTsaTrust(
  check: TsaCheck,
  trustList: string[],
): Promise<TsaTrustResult> {
  if (!check.tsaPresent || check.embeddedCertCount === 0) {
    return {
      trusted: null,
      reason: "TSA not present or no embedded cert; nothing to validate.",
      signerSubject: check.signerSubject ?? undefined,
      signerIssuer: check.signerIssuer ?? undefined,
    };
  }
  if (trustList.length === 0) {
    return {
      trusted: null,
      reason: "Empty trust list — chain-to-root check skipped. Pass roots to enforce.",
      signerSubject: check.signerSubject ?? undefined,
      signerIssuer: check.signerIssuer ?? undefined,
    };
  }
  configureEngine();
  for (const pem of trustList) {
    let rootCert: Certificate;
    try {
      rootCert = parsePemCert(pem);
    } catch (e) {
      return {
        trusted: false,
        reason: `Trust list entry failed to parse: ${describeError(e)}`,
      };
    }
    const rootSubject = describeDn(rootCert.subject.typesAndValues as DnTypeAndValue[]);
    if (rootSubject === check.signerIssuer) {
      return {
        trusted: true,
        reason: `Signer issuer matches pinned root (${rootSubject}).`,
        signerSubject: check.signerSubject ?? undefined,
        signerIssuer: check.signerIssuer ?? undefined,
      };
    }
  }
  return {
    trusted: false,
    reason: `Signer issuer "${check.signerIssuer}" does not match any pinned root.`,
    signerSubject: check.signerSubject ?? undefined,
    signerIssuer: check.signerIssuer ?? undefined,
  };
}

/* ------- helpers ------- */

function parsePemCert(pem: string): Certificate {
  const lines = pem.replace(/\r/g, "").split("\n").map((l) => l.trim());
  const b64 = lines.filter((l) => l && !l.startsWith("-----")).join("");
  const der = decodeBase64(b64);
  const asn = asn1js.fromBER(der.slice().buffer);
  if (asn.offset === -1) {
    throw new Error("PEM contents did not parse as ASN.1.");
  }
  return new Certificate({ schema: asn.result });
}

type DnTypeAndValue = {
  type: string;
  value: { valueBlock: { value: string } };
};

function describeDn(entries: DnTypeAndValue[]): string {
  const map: Record<string, string> = {
    "2.5.4.3": "CN",
    "2.5.4.6": "C",
    "2.5.4.7": "L",
    "2.5.4.8": "ST",
    "2.5.4.10": "O",
    "2.5.4.11": "OU",
    "1.2.840.113549.1.9.1": "emailAddress",
  };
  return entries
    .map((e) => {
      const label = map[e.type] ?? e.type;
      const v = e.value?.valueBlock?.value ?? "";
      return `${label}=${v}`;
    })
    .join(",");
}

function describeError(e: unknown): string {
  if (e instanceof Error) return e.message;
  return String(e);
}

function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a[i]! ^ b[i]!;
  return diff === 0;
}

File: lib-python/eatf_verifier/tsa.py
"""RFC 3161 timestamp parsing + verification.

Mirrors lib/src/tsa.ts. The TS reference uses pkijs; the Python
port uses asn1crypto for the TimeStampToken / CMS structure and
pyca/cryptography for the SignerInfo signature verification.

This is "structural-check + SignerInfo signature only" parity
with the TS reference. Chain-to-root validation against pinned
TSA roots is documented as planned in lib-python/README.md.
"""

from __future__ import annotations

import base64
from dataclasses import dataclass

from asn1crypto import cms, tsp
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding, rsa
from cryptography.x509 import load_der_x509_certificate


@dataclass
class TsaCheck:
    tsa_present: bool
    raw_size_bytes: int
    gen_time: str | None  # ISO 8601 string, or None
    message_imprint_matches: bool | None  # None when we can't determine
    signature_verified: bool | None  # None when no embedded cert
    signer_subject: str | None
    signer_issuer: str | None
    embedded_cert_count: int = 0


@dataclass
class TsaTrustResult:
    """Outcome of the trust-anchor cross-check.

    `trusted=True`  — TSA signing cert's issuer matches a pinned root's subject.
    `trusted=False` — no pinned root matched.
    `trusted=None`  — either no embedded cert in the TSA token, or
                       an empty trust list was supplied (check skipped).
    """

    trusted: bool | None
    reason: str
    signer_subject: str | None = None
    signer_issuer: str | None = None


def inspect_tsa(tsr_b64: str, expected_hash_hex: str) -> TsaCheck:
    """Parse and structurally verify an RFC 3161 token.

    `tsr_b64` is the contents of timestamp.tsr (base64) — the
    verifier strips ASCII whitespace and decodes. `expected_hash_hex`
    is the hex SHA-256 the verifier expects to find as the message
    imprint.
    """
    try:
        raw = base64.b64decode(tsr_b64.strip(), validate=False)
    except Exception:
        return _empty()

    if not raw:
        return _empty()

    # An RFC 3161 response can be either:
    #   1. A bare TimeStampResp (with status + timeStampToken)
    #   2. A bare TimeStampToken (CMS SignedData)
    # We accept both shapes.
    token_bytes = _extract_token(raw)
    if token_bytes is None:
        return _empty()

    try:
        signed = cms.ContentInfo.load(token_bytes)
        if signed["content_type"].native != "signed_data":
            return _empty()
        sd = signed["content"]
    except Exception:
        return _empty()

    try:
        tst_info_bytes = sd["encap_content_info"]["content"].native
        tst_info = tsp.TSTInfo.load(tst_info_bytes)
        imprint = tst_info["message_imprint"]["hashed_message"].native
        imprint_hex = imprint.hex()
        message_imprint_matches = imprint_hex == expected_hash_hex.lower()
        gen_time = tst_info["gen_time"].native.isoformat() if tst_info["gen_time"].native else None
    except Exception:
        message_imprint_matches = None
        gen_time = None

    signer_subject = None
    signer_issuer = None
    signature_verified: bool | None = None
    embedded_cert_count = 0

    try:
        certs = sd["certificates"]
        if certs:
            embedded_cert_count = len(certs)
            cert_der = certs[0].chosen.dump()
            cert = load_der_x509_certificate(cert_der)
            signer_subject = cert.subject.rfc4514_string()
            signer_issuer = cert.issuer.rfc4514_string()

            # Verify the first SignerInfo against this cert.
            signer_info = sd["signer_infos"][0]
            signature = signer_info["signature"].native

            # Compute the signed-attributes DER (or fall back to
            # encap_content_info content).
            if signer_info["signed_attrs"]:
                signed_attrs = signer_info["signed_attrs"]
                signed_attrs_der = signed_attrs.dump(force=True)
                # CMS requires the signed-attributes to be re-encoded
                # with the SET OF tag (0x31), not the implicit [0]
                # context-specific tag they're carried with on the wire.
                signed_attrs_der = b"\x31" + signed_attrs_der[1:]
                tbs = signed_attrs_der
            else:
                tbs = tst_info_bytes

            pubkey = cert.public_key()
            if isinstance(pubkey, rsa.RSAPublicKey):
                try:
                    pubkey.verify(signature, tbs, padding.PKCS1v15(), hashes.SHA256())
                    signature_verified = True
                except InvalidSignature:
                    signature_verified = False
            else:
                # ECDSA TSA signing is rare in v0.1 deployments;
                # treat as inconclusive for now (verifier ignores
                # signature_verified=None).
                signature_verified = None
    except Exception:
        # Parsing failure → leave signature_verified at None
        pass

    return TsaCheck(
        tsa_present=True,
        raw_size_bytes=len(raw),
        gen_time=gen_time,
        message_imprint_matches=message_imprint_matches,
        signature_verified=signature_verified,
        signer_subject=signer_subject,
        signer_issuer=signer_issuer,
        embedded_cert_count=embedded_cert_count,
    )


def verify_tsa_trust(
    check: TsaCheck, trust_list: list[bytes]
) -> TsaTrustResult:
    """Mirror of lib/src/tsa.ts: verifyTsaTrust.

    Decision rule: the TSA signing cert's *issuer DN* must equal
    one of the trust list's *subject DN*s. This is a single-step
    pin check, not full RFC 5280 path validation.

    Returns `trusted=None` when there's nothing to validate
    (no embedded cert, or empty trust list).
    """
    if not check.tsa_present or check.embedded_cert_count == 0:
        return TsaTrustResult(
            trusted=None,
            reason="TSA not present or no embedded cert; nothing to validate.",
            signer_subject=check.signer_subject,
            signer_issuer=check.signer_issuer,
        )
    if not trust_list:
        return TsaTrustResult(
            trusted=None,
            reason="Empty trust list — chain-to-root check skipped. Pass roots to enforce.",
            signer_subject=check.signer_subject,
            signer_issuer=check.signer_issuer,
        )
    for pem in trust_list:
        try:
            root_cert = _parse_pem_cert(pem)
        except Exception as e:
            return TsaTrustResult(
                trusted=False,
                reason=f"Trust list entry failed to parse: {e}",
            )
        root_subject = root_cert.subject.rfc4514_string()
        if root_subject == check.signer_issuer:
            return TsaTrustResult(
                trusted=True,
                reason=f"Signer issuer matches pinned root ({root_subject}).",
                signer_subject=check.signer_subject,
                signer_issuer=check.signer_issuer,
            )
    return TsaTrustResult(
        trusted=False,
        reason=(
            f'Signer issuer "{check.signer_issuer}" does not match any pinned root.'
        ),
        signer_subject=check.signer_subject,
        signer_issuer=check.signer_issuer,
    )


def _parse_pem_cert(pem: bytes):
    """Decode a PEM CERTIFICATE block into a cryptography x509 object."""
    import base64

    text = pem.decode("ascii", errors="replace")
    body = "".join(
        line for line in text.splitlines() if not line.startswith("-----")
    )
    der = base64.b64decode(body)
    return load_der_x509_certificate(der)


def _extract_token(raw: bytes) -> bytes | None:
    """Try parse as TimeStampResp first; fall back to bare token."""
    try:
        resp = tsp.TimeStampResp.load(raw)
        # Status must be 0 (granted) or 1 (granted with mods)
        status = resp["status"]["status"].native
        if status not in ("granted", "granted_with_mods"):
            return None
        if resp["time_stamp_token"]:
            return resp["time_stamp_token"].dump()
    except Exception:
        pass

    # Try as bare CMS ContentInfo
    try:
        cms.ContentInfo.load(raw)
        return raw
    except Exception:
        return None


def _empty() -> TsaCheck:
    return TsaCheck(
        tsa_present=False,
        raw_size_bytes=0,
        gen_time=None,
        message_imprint_matches=None,
        signature_verified=None,
        signer_subject=None,
        signer_issuer=None,
        embedded_cert_count=0,
    )
